### ISSUE TITLE
Illegal memory access when running large models on GPU

### ISSUE THREAD

**Issue by PershingSquare** [tensorflow/tensorflow#44318]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Linux RedHat 7.6**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): **Docker container**
- TensorFlow version (use command below): **I used the one from the tag 2.3.0-gpu**
- Python version: **3.6.8**
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: **10.1.243**
- GPU model and memory: **Tesla V100-SXM2-32GB**

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When I run a large model, it I get an illegal memory access was encountered. This occurs even when I use managed memory.

**Describe the expected behavior**
I expect to see either an OOM error or no error when I used managed memory to emulate a larger memory size.
**Standalone code to reproduce the issue**

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

     import tensorflow as tf 
     import numpy as np
     # Uncomment this to use CUDA Managed Memory (This issue still happens regardless if this is uncommented or not).
     #config = tf.compat.v1.ConfigProto()
     #config.gpu_options.per_process_gpu_memory_fraction = 5 
     #session = tf.compat.v1.InteractiveSession(config=config)
 
     # Grab some random dataset
     fashion_mnist = tf.keras.datasets.fashion_mnist
     (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
 
     # Standard MNIST Example, except with a large model
     model = tf.keras.Sequential([
         tf.keras.layers.Flatten(input_shape=(28, 28)),
         tf.keras.layers.Dense(2048 * 2048, activation='relu'),
         tf.keras.layers.Dense(10)
      ])
 
     model.compile(optimizer='adam',
             loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
             metrics=['accuracy'])
 
     model.fit(train_images, train_labels, epochs=1)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
Here is full log
[tf.log](https://github.com/tensorflow/tensorflow/files/5436897/tf.log)
This part seems relevant.
     Non-OK-status: GpuLaunchKernel(FillPhiloxRandomKernelLaunch<Distribution>, num_blocks, block_size, 0, d.stream(), gen, data, size, dist) status: Internal: an illegal memory access was encountered

**Comment by amahendrakar** [comment]
@PershingSquare,
I was able to reproduce the issue using the given code snippet. 

However, setting a hard limit on the total GPU memory as shown in the guide [here](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth), I ran into an OOM error as expected. Please check [this gist](https://colab.research.google.com/gist/amahendrakar/b9425771f853d3ac74cf60de8efcaeb9/44318.ipynb) for reference. Thanks!

**Comment by PershingSquare** [comment]
What causes the GpuLaunchKernel(FillPhiloxRandomKernelLaunch, num_blocks, block_size, 0, d.stream(), gen, data, size, dist) status: Internal: an illegal memory access was encountered
portion and how I do avoid it?

**Comment by amahendrakar** [comment]
@PershingSquare,
>`tf.keras.layers.Dense(2048 * 2048, activation='relu')`

Looks like the error was caused due to the size of the Dense layer. You can avoid this by limiting the GPU memory as mentioned in [this guide](per_process_gpu_memory_fraction). Thanks!
